spec/lang/representation.md

```diff
@@ -123,19 +125,72 @@ fn encode_ptr<M: Memory>(ptr: ThinPointer<M::Provenance>) -> List<AbstractByte<M
     
 }
 
+impl PtrType {
+    // Returns a 2-tuple type representing this wide pointer or `None` if it is thin.
+    fn as_tuple_ty<M: Memory>(self) -> Option<Type> {
+        let meta_ty = match self.meta_kind() {
+            PointerMetaKind::None => throw!(),
+            PointerMetaKind::ElementCount => Type::Int(IntType { signed: Unsigned, size: M::T::PTR_SIZE }),
+        };
+        let thin_pointer_field = (Offset::ZERO, Type::Ptr(PtrType::Raw { meta_kind: PointerMetaKind::None }));
+        let metadata_field = (Offset::from_bytes(M::T::PTR_SIZE.bytes()).expect("PTR_SIZE is positive"), meta_ty);
+        ret(Type::Tuple {
+            fields: list![thin_pointer_field, metadata_field],
+            size: Int::from(2) * M::T::PTR_SIZE,
+            align: Align::from_bytes(M::T::PTR_SIZE.bytes()).expect("PTR_SIZE is aligned"),
+        })
+    }
+}
+
+impl PointerMeta {
+    fn from<M: Memory>(value: Value<M>) -> Option<PointerMeta> {
+        match value {
+            Value::Int(count) => Some(PointerMeta::ElementCount(count)),
+            _ => None,
+        }
+    }
+
+    fn into_value<M: Memory>(self) -> Value<M> {
+        match self {
+            PointerMeta::ElementCount(count) => Value::Int(count),
+        }
+    }
+}
+
 impl Type {
     fn decode<M: Memory>(Type::Ptr(ptr_type): Self, bytes: List<AbstractByte<M::Provenance>>) -> Option<Value<M>> {
-        // TODO(UnsizedTypes): Decode pointer metadata
-        let ptr = decode_ptr::<M>(bytes)?;
-        if !ptr_type.addr_valid(ptr.addr) {
-            throw!();
+        if ptr_type.meta_kind() == PointerMetaKind::None {
+            let ptr = decode_ptr::<M>(bytes)?;
+            if !ptr_type.addr_valid(ptr.addr) {
+                throw!();
+            }
+            ret(Value::Ptr(ptr.widen(None)))
+        } else {
+            let tuple_ty = ptr_type.as_tuple_ty::<M>().expect("ptr_type has metadata");
+            let Value::Tuple(parts) = tuple_ty.decode::<M>(bytes)? else {
+                panic!("as_tuple_ty always returns a tuple type");
+            };
+            let Value::Ptr(ptr) = parts[0] else {
+                panic!("as_tuple_ty always returns tuple with the first field being a thin pointer");
+            };
+            let meta = PointerMeta::from(parts[1]);
```

Comment:
Shouldn't this panic if we get a `None`, since as_tuple_ty should always produce a suitable tuple type?